runtime/fastly/host-api/host_api.cpp

```diff
@@ -196,6 +198,187 @@ Response make_response(fastly::fastly_host_http_response &resp) {
   return Response{HttpResp{resp.f0}, HttpBody{resp.f1}};
 }
 
+template <auto header_names_get>
+Result<std::vector<HostString>> generic_get_header_names(auto handle) {
+  Result<std::vector<HostString>> res;
+
+  fastly::fastly_world_list_string ret;
+  fastly::fastly_host_error err;
+  fastly::fastly_world_string *strs = static_cast<fastly::fastly_world_string *>(
+      cabi_malloc(LIST_ALLOC_SIZE * sizeof(fastly::fastly_world_string), 1));
+  size_t str_max = LIST_ALLOC_SIZE;
+  size_t str_cnt = 0;
+  size_t nwritten;
+  uint8_t *buf = static_cast<uint8_t *>(cabi_malloc(HOSTCALL_BUFFER_LEN, 1));
+  uint32_t cursor = 0;
+  int64_t next_cursor = 0;
+  while (true) {
+    if (!convert_result(
+            header_names_get(handle, buf, HEADER_MAX_LEN, cursor, &next_cursor, &nwritten), &err)) {
+      cabi_free(buf);
+      res.emplace_err(err);
+      return res;
+    }
+    if (nwritten == 0) {
+      break;
+    }
+    uint32_t offset = 0;
+    for (size_t i = 0; i < nwritten; i++) {
+      if (buf[i] != '\0')
+        continue;
+      if (str_cnt == str_max) {
+        strs = static_cast<fastly::fastly_world_string *>(
+            cabi_realloc(strs, str_max * sizeof(fastly::fastly_world_string), 1,
+                         (str_max + LIST_ALLOC_SIZE) * sizeof(fastly::fastly_world_string)));
+        str_max += LIST_ALLOC_SIZE;
+      }
+      strs[str_cnt].ptr = static_cast<uint8_t *>(cabi_malloc(i - offset + 1, 1));
+      strs[str_cnt].len = i - offset;
+      memcpy(strs[str_cnt].ptr, buf + offset, i - offset + 1);
```

Comment:
Could you compute the length separately and use it for both the allocation size and `len` field? It looks like there's one extra char allocated than is needed right now, as I'm pretty sure it's not required that a `HostString` be null-terminated.

Reply:
This was in the host API work Jake did, just moved around, but yeah I can remove the unused null byte allocation.